proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,262 @@
+validator. This results in compilation passing but the validator failing. For
+example take the following code:
+
+> [Godbolt Link](https://godbolt.org/z/v1sjEEETW)
+```c++
+Texture2D texture : register(t0);
+SamplerState samplerState : register(s0);
+
+FeedbackTexture2D<SAMPLER_FEEDBACK_MIN_MIP> map : register(u0);
+
+[numthreads(4, 4, 1)] 
+void main(uint3 threadId : SV_DispatchThreadId) {
+  float2 uv = threadId.xy;
+  uv /= 256;
+
+  map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+}
+```
+
+Compiled with the `-T cs_6_4` flag this produces the validation error:
+```
+<>:11:3: error: Opcode WriteSamplerFeedbackLevel not valid in shader model cs_6_4.
+note: at 'call void @dx.op.writeSamplerFeedbackLevel(i32 176, %dx.types.Handle %1, %dx.types.Handle %2, %dx.types.Handle %3, float %8, float %9, float undef, float undef, float %11)' in block '#0' of function 'main'.
+Validation failed.
+```
+
+This error displays the LLVM IR text for the error site. In a trivial example
+like the one above finding the error is easy. In a complex shader that could be
+thousands of lines long this is unmanageable.
+
+With this proposal under the _relaxed_ mode, clang will emit the following
+warning:
+
+```
+<>:11:6: warning: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   11 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+With this proposal under the _default_ or _strict_ mode, clang will emit the
+following error:
+
+```
+<>:11:6: error: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   11 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+To illustrate the difference between _default_ and _strict_ take the following
+example:
+
+```c++
+Texture2D texture : register(t0);
+SamplerState samplerState : register(s0);
+
+FeedbackTexture2D<SAMPLER_FEEDBACK_MIN_MIP> map : register(u0);
+
+void fn(uint3 threadId) {
+  float2 uv = threadId.xy;
+  uv /= 256;
+  map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+}
+
+[numthreads(4, 4, 1)] 
+void main(uint3 threadId : SV_DispatchThreadId) { }
+```
+
+In this example the call of the unavailable API is in an unused function. In
+both the _relaxed_ and _default_ diagnostic modes, clang will emit no
+diagnostics. In the _strict_ mode, clang emits the following diagnostic:
+
+```
+<>:9:6: error: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   9 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+     |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+In this case the DXIL validator will emit no diagnostic since the call to the
+unavailable API is not present in the final output.
+
+In all three cases above the DXIL validator will also emit the same diagnostic
+as it does today.

```

Comment:
This is confusing as it follows the sentence "In this case the DXIL validator will emit no diagnostic".